src/Storages/MergeTree/Compaction/MergeSelectors/SimpleMergeSelector.cpp

```diff
@@ -86,9 +77,47 @@ struct Estimator
         return (sum_size + sum_size_fixed_cost * count) / (count - 1.9);
     }
 
-    double min_score = 0.0;
-    Iterator best_begin {};
-    Iterator best_end {};
+    void prepare(size_t count)
+    {
+        count = std::min(count, ranges.size());
+        std::partial_sort(ranges.rbegin(), ranges.rbegin() + count, ranges.rend(), [](const ScoredRange & lhs, const ScoredRange & rhs)
+        {
+            return lhs.score < rhs.score;
+        });
+        sorted_suffix_size = count;
+    }
+
+    std::optional<PartsRange> buildMergeRange(size_t max_total_size_to_merge)
+    {
+        while (!ranges.empty())
+        {
+            if (sorted_suffix_size == 0)
+                prepare(ranges.size());
```

Comment:
It will sort the whole `ranges` vector. Maybe better to sort more `count` elements if it is less than `ranges.size()`? If all ranges are in the worst case O(N^2) then sorting all the ranges will be O(N^2 * log (N)) in time. This does not seem fast enough.